### ISSUE TITLE
NamedTuple Subclass Thinking It Returns a Tuple

### ISSUE THREAD

**Issue by Py-mon** [microsoft/pyright#9147]
## Environment data

-   Language Server version: v2024.6.1
-   OS and version: Windows 11
-   Python version (& distribution if applicable, e.g. Anaconda): 3.11

## Code Snippet

```python
from typing import NamedTuple


class Vector(NamedTuple("Vector", [("y", int), ("x", int)])):
    def __add__(self, vector: tuple[int, int]):
        cls = type(self)
        vector = super().__new__(cls, *vector)
        return type(self)(self.y + vector.y, self.x + vector.x)


print(Vector(1, 2) + (3, 1))
```
![image](https://github.com/microsoft/pyright/assets/102424561/c098e105-2c06-4c86-9093-6e0169ee3910)

## Expected behavior

When hovering over vector, it is the type `Vector`

## Actual behavior

When hovering over vector, it is the type `tuple[int, int]`

## Logs

```
2024-06-25 19:40:40.068 [info] Send text to terminal: 
2024-06-25 19:40:40.277 [info] Send text to terminal: cls
2024-06-25 19:40:40.277 [info] Send text to terminal:  c:; cd 'c:\Users\jacob\OneDrive\Documents\Programming\Python\Pyframe'; & 'c:\Users\jacob\AppData\Local\Programs\Python\Python311\python.exe' 'c:\Users\jacob\.vscode\extensions\ms-python.debugpy-2024.6.0-win32-x64\bundled\libs\debugpy\adapter/../..\debugpy\launcher' '55472' '--' 'C:\Users\jacob\OneDrive\Documents\Programming\Python\Pyframe\rep.py' 
```

**Comment by erictraut** [comment]
Yes, the type of `vector` should be narrowed on assignment in this case. Pyright wasn't handling this narrowing case correctly because the narrowed type is a type variable (`Self`) rather than a concrete type.

This will be fixed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.384
